refactor datetime output

nilsnolde opened the pull request:
Route & matrix didn't share the code to generate the response's datetime strings. Can't remember why I didn't do that in https://github.com/valhalla/valhalla/pull/4071. I thought I'll be smarter with my PRs for once, and not cram refactors together with functional changes:) Comes from the experience of reviewing such PRs I guess (not pointing fingers!). 

Really, it's to prepare for outputting timezone-aware datetime strings: https://github.com/valhalla/valhalla/issues/4241. Last chance to tell me that no one wants that:) It's a breaking change the way I'd do it (i..e. just return `2023-02-27T22:50+02:00` instead of `2023-02-27T22:50`). Depending on the client code that could be a problem.. IMO it's acceptable.

EDIT: Also fixes #4360 as collateral damage..

nilsnolde commented on src/baldr/datetime.cc:
```diff
@@ -524,6 +524,14 @@ uint32_t second_of_week(uint32_t epoch_time, const date::time_zone* time_zone) {
   return day * midgard::kSecondsPerDay + since_midnight.count();
 }
 
+std::string
+offset_date(const std::string& in_dt, const uint32_t in_tz, const uint32_t out_tz, float offset) {
+  // get the input UTC time, add the offset and translate to the out timezone
+  auto iepoch = DateTime::seconds_since_epoch(in_dt, DateTime::get_tz_db().from_index(in_tz));
+  auto oepoch =
+      static_cast<uint64_t>(static_cast<double>(iepoch) + static_cast<double>(offset + .5f));
+  return DateTime::seconds_to_date(oepoch, DateTime::get_tz_db().from_index(out_tz), false);
+}
```
I broke it up a bit, so we can put this here in baldr/date_time without including GraphReader too. I added a function to GraphReader instead.

nilsnolde commented on src/baldr/graphreader.cc:
```diff
@@ -975,6 +975,16 @@ int GraphReader::GetTimezone(const baldr::GraphId& node, graph_tile_ptr& tile) {
   return (tile == nullptr) ? 0 : tile->node(node)->timezone();
 }
 
+int GraphReader::GetTimezoneFromEdge(const baldr::GraphId& edge, graph_tile_ptr& tile) {
+  auto nodes = GetDirectedEdgeNodes(edge, tile);
+  if (const auto* node = nodeinfo(nodes.first, tile))
+    return node->timezone();
+  else if (const auto* node = nodeinfo(nodes.second, tile))
+    return node->timezone();
+
+  return 0;
+}
```
analogous to GraphReader::GetTimezone for nodes

nilsnolde commented on src/thor/costmatrix.cc:
```diff
@@ -233,23 +234,31 @@ void CostMatrix::SourceToTarget(Api& request,
-                                   time_infos[origin_idx].timezone_index,
-                                   target_edgelabel_[target_idx].front().edgeid(), graphreader,
-                                   static_cast<uint64_t>(time));
+    float time = connection.cost.secs;
```
adding `0.5f` makes no sense here

nilsnolde replied:
in fact this was a bug some clients were complaining about, where matrix had a one second offset to /route. but not always. now I know why haha

nilsnolde commented on src/thor/costmatrix.cc:
```diff
@@ -233,23 +234,31 @@ void CostMatrix::SourceToTarget(Api& request,
-                                   static_cast<uint64_t>(time));
+    float time = connection.cost.secs;
+    // no datetime for same locations or unfound connections
+    if (time && time < kMaxCost) {
```
only add date_time if it's not 0 and the connection was found

nilsnolde commented on src/thor/costmatrix.cc:
```diff
@@ -233,23 +234,31 @@ void CostMatrix::SourceToTarget(Api& request,
-                                   static_cast<uint64_t>(time));
+    float time = connection.cost.secs;
+    // no datetime for unfound connections
+    if (time < kMaxCost) {
```
we don't add date_time when there's no connection, but we do add it for origin == target. in the serializer we put `null` when there was no connection

nilsnolde commented on src/tyr/matrix_serializer.cc:
```diff
@@ -12,6 +12,24 @@ using namespace valhalla::thor;
 
 namespace {
 
+json::ArrayPtr serialize_datetime(const valhalla::Matrix& matrix,
```
I realized there was no date_time array for a "slim" matrix. The change is contained only in this file (+ test), so I figured it's fine to cram in here

nilsnolde commented on src/tyr/matrix_serializer.cc:
```diff
@@ -150,16 +168,24 @@ std::string serialize(const Api& request, double distance_scale) {
     }
     matrix->emplace("distances", distance);
     matrix->emplace("durations", time);
+    if (has_dt) {
+      matrix->emplace("date_times", date_time);
+    }
```
we only put the date_times array if at least one connection had a date_time

nilsnolde commented on valhalla/thor/matrix_common.h:
```diff
@@ -56,36 +56,6 @@ struct Destination {
-  if (pred_id.Is_Valid()) {
-    // get the timezone of the output location
-    auto out_nodes = reader.GetDirectedEdgeNodes(pred_id, tile);
-    dest_tz = reader.GetTimezone(out_nodes.first, tile) || reader.GetTimezone(out_nodes.second, tile);
```
I just realized, this PR will also fix #4360 . this line is not what python-eyes think it is.. `dest_tz` will be true or false, i.e. 1 or 0. And 1 must be UTC. The new logic doesn't have that flaw.

nilsnolde commented:
I drafted this bcs we agreed to output the timezone info as an extra response field. I’d just put the „+02:00“ in an extra field.

kevinkreiser commented on valhalla/baldr/graphreader.h:
```diff
@@ -947,6 +947,15 @@ class GraphReader {
+   * @param tile   Current tile.
+   * @return Returns the timezone index. A value of 0 indicates an invalid timezone.
+   */
+  int GetTimezoneFromEdge(const baldr::GraphId& edge, graph_tile_ptr& tile);
```
just need to check and mention if the tile is modified. if its not modified we should pass it wihtout the reference if modified we should say whether its for the end or begin node of the edge

nilsnolde replied:
you're right, in `nodeinfo()` the tile can change, so we could end up with the edge's end node's tile instead. is a comment in the docstring good enough for you or you want smth more elaborate?

kevinkreiser replied:
yeah just docstring update was all i wanted. i hope that its a gaurantee though that as a post condition it will be the endnodes tile.

nilsnolde replied:
Fwiw, our own usage of this (so far) is safe, we don’t even use the tile, only for caching

kevinkreiser replied:
yep yep its just to avoid mistakes

nilsnolde replied:
> i hope that its a gaurantee though that as a post condition it will be the endnodes tile not just could be, i guess when the tile isnt found it could be, thats ok.

after this function is called, it's very unlikely that the tile is changed (unless you compiled the graph without timezone info, but maybe that's even handled further up in the processing chain). that only happens if the edge's start node has no timezone info.

kevinkreiser commented:
the history and subsequently the diff of this branch are screwed up.

nilsnolde commented:
fixed. btw, regarding your comment if the tile is being switched in the new `GetTimezoneFromEdge` function, I added that as info to the docstring. is that good enough for you?

nilsnolde commented:
@chrstnbwnkl will PR tmrw or so with the next bit to actually return the tz offset in the response:)
